### PR TITLE
[SYCL] Fix `static|dynamic_address_cast` to `generic`

### PR DESCRIPTION
**Pull request by aelovikov-intel**
SPIRV operations are defined such that `OpGenericCastToPtr` and
`OpGenericCastToPtrExplicit` cannot be used when target `Storage Class`
is `Generic`, yet we were generating such code. This PR fixes that.

### CODE REVIEW

File: sycl/test/check_device_code/extensions/address_cast.cpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+// CHECK-NEXT:    ret void
+//
+SYCL_EXTERNAL auto to_global_not_decorated(int *p) {
+  return static_address_cast<access::address_space::global_space,
+                             access::decorated::no>(p);
+}
+// CHECK-LABEL: define dso_local spir_func void @_ZN14static_as_cast20to_generic_decoratedEPi(
+// CHECK-SAME: ptr addrspace(4) dead_on_unwind noalias nocapture writable writeonly sret(%"class.sycl::_V1::multi_ptr.1") align 8 [[AGG_RESULT:%.*]], ptr addrspace(4) noundef [[P:%.*]]) local_unnamed_addr #[[ATTR2:[0-9]+]] !srcloc [[META22:![0-9]+]] !sycl_fixed_targets [[META7]] {
+// CHECK-NEXT:  [[ENTRY:.*:]]
+// CHECK-NEXT:    store ptr addrspace(4) [[P]], ptr addrspace(4) [[AGG_RESULT]], align 8, !tbaa [[TBAA23:![0-9]+]], !alias.scope [[META25:![0-9]+]]
```
Comment (aelovikov-intel) [inline sycl/test/check_device_code/extensions/address_cast.cpp:40]: Use per-commit diff to see how this PR affects code generation here. First commit is test-only without functional changes.

**Review by againull — APPROVED** [review]

File: sycl/include/sycl/ext/oneapi/experimental/address_cast.hpp [hunk 2]
```diff
@@ -22,8 +22,19 @@ template <access::address_space Space, access::decorated DecorateAddress,
 multi_ptr<ElementType, Space, DecorateAddress>
 static_address_cast(ElementType *Ptr) {
 #ifdef __SYCL_DEVICE_ONLY__
-  auto CastPtr = sycl::detail::spirv::GenericCastToPtr<Space>(Ptr);
-  return multi_ptr<ElementType, Space, DecorateAddress>(CastPtr);
+  // TODO: Remove this restriction.
+  static_assert(std::is_same_v<ElementType, remove_decoration_t<ElementType>>,
+                "The extension expect undecorated raw pointers only!");
```
Comment (Pennycook) [inline sycl/include/sycl/ext/oneapi/experimental/address_cast.hpp:27]: Out of curiosity: is this something that would be difficult to fix?  Why not just use `remove_decoration_t` to create an undecorated raw pointer here, and then use the undecorated pointer for the rest of the function?

**Review by Pennycook — APPROVED** [review]
LGTM, but I do have a question.

File: sycl/include/sycl/ext/oneapi/experimental/address_cast.hpp (hunk 2 above)
Comment (aelovikov-intel) [reply sycl/include/sycl/ext/oneapi/experimental/address_cast.hpp:27]: I plan to do that in the next PR(s). It has to come with the doc change as well.

**Pull request merged**